protocol/test-infra.md

```diff
@@ -0,0 +1,358 @@
+When there are multiple alternatives there should be an explanation
+of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API)
+is likely too low level. -->
+
+## Towards a solution
+
+Based on the challenges outlined above, I believe we need:
+- Automation to spin up test chains
+- Flexibility with node types
+- Separation of setup and test
+- More expressive language to define tests in
+- Improved insights
+
+### Automation to spin up test chains
+
+The automation story is largely a platforms/infra project.
+With `op-deployer` and Kurtosis as the main solution to setup configs and nodes quickly.
+
+### Flexibility with node types
+
+Flexibility requires some standardization:
+each execution-engine talks the same RPC,
+but rollup-nodes, peripherals, alternative proof systems, etc. still have unique API interfaces.
+
+Defining some interfaces (especially RPC namespaces) as "standard", can go a long way for portable testing.
+
+### Separation of setup and test
+
+One pattern we are trying to work towards in `op-e2e` is to hide
+the test-setup and node-access with Go interfaces as much as possible.
+The more it is separated by interface, the more flexible the implementation of the setup is,
+and the more reusable the test.
+
+One example of this is the Go testing that also runs against local docker-compose devnet,
+which was unfortunately removed in [PR 12216](https://github.com/ethereum-optimism/optimism/pull/12216).
+
+### More expressive language to define tests in
+
+A test should be simple, easy to maintain by anyone.
+
+At the same time, a test should be able to navigate into an edge-case,
+to cover the more subtle state-transition problems.
+
+A lot of the time, the test is complicated due to the setup,
+and not having common idioms to access nodes and state that assertions are made over.
+
+Besides separating from the setup, we need to review how we express the tests themselves.
+Part of this is the language itself, part of it is the DSL (domain specific language) we build on top.
+
+#### Golang
+
+Golang can be great:
+- A lot of the existing tests and test-utilities can be ported over quickly.
+- There is a Kurtosis Go SDK we can integrate with
+- Golang is the simplest common denominator for test-writing (assuming no introduction of Python just for testing).
+
+While there are many existing test utilities and abstractions, the Go test stack can still improve.
+Notably it can be challenging to fork a test into two independent tests, or parametrize in general.
+
+With these existing test utils, some of it already looks like a DSL.
+E.g. the action-tests have a specific way of making one thing happen at a time, in a bigger orchestration of actors.
+
+This DSL-like functionality needs to improve in the system-tests,
+where awaiting events and more asynchronous work can still feel verbose and fragile.
+
+#### Rust
```

Comment:
Strongly in favor of writing tests in Go. I think adopting Rust here will make things much slower, and force us to rewrite all of our existing tooling.